src/gpl/src/cudasplibs.cu

```diff
@@ -0,0 +1,119 @@
+#include "cudasplibs.h"
+#include "utl/Logger.h"
+
+void cudasplibs::cudaerror(cudaError_t code) {
+    if (code != cudaSuccess){
+        log_->report("[CUDA ERROR] {} at line {} in file {} \n", cudaGetErrorString(code), __LINE__, __FILE__);
+        cudaDeviceReset();
+    }
+}
+void cudasplibs::cusparseerror(cusparseStatus_t code) {
+    if (code != CUSPARSE_STATUS_SUCCESS){
+        log_->report("[CUSPARSE ERROR] {} at line {} in file {}\n", cusparseGetErrorString(code), __LINE__, __FILE__); 
+        cudaDeviceReset();
+    }
+}
+
+void cudasplibs::cusolvererror(cusolverStatus_t code) {
+    if (code != CUSOLVER_STATUS_SUCCESS){
+        log_->report("[CUSOLVER ERROR] {} at line {} in file {}\n", cudaGetErrorString(*(cudaError_t*)&code), __LINE__, __FILE__); 
+        cudaDeviceReset();
+    }
+}
+
+cudasplibs::cudasplibs(std::vector<int>& cooRowIndex, std::vector<int>& cooColIndex, std::vector<float>& cooVal, Eigen::VectorXf& fixedInstForceVec, utl::Logger* logger){
+    m = fixedInstForceVec.size();
+    nnz = cooVal.size();
+    log_ = logger;
+    // Allocate device memeory and copy data to device
+    cudaerror(cudaMalloc((void**)&d_cooRowIndex, nnz * sizeof(int)));
+    cudaerror(cudaMalloc((void**)&d_cooColIndex, nnz * sizeof(int)));
+    cudaerror(cudaMalloc((void**)&d_cooVal, nnz * sizeof(float)));
+    cudaerror(cudaMalloc((void**)&d_fixedInstForceVec, m * sizeof(float)));
+    cudaerror(cudaMalloc((void**)&d_instLocVec, m * sizeof(float)));
+          
+    // Copy data (COO storage method)
+    cudaerror(cudaMemcpy(d_cooRowIndex, cooRowIndex.data(), sizeof(int)*nnz, cudaMemcpyHostToDevice));
+    cudaerror(cudaMemcpy(d_cooColIndex, cooColIndex.data(), sizeof(int)*nnz, cudaMemcpyHostToDevice));
+    cudaerror(cudaMemcpy(d_cooVal, cooVal.data(), sizeof(float)*nnz, cudaMemcpyHostToDevice));
+    cudaerror(cudaMemcpy(d_fixedInstForceVec, fixedInstForceVec.data(), sizeof(float)*m, cudaMemcpyHostToDevice));
+}
+
+void cudasplibs::cusolverSpQR(Eigen::VectorXf& instLocVec){
+    // Parameters that don't change with iteration and used in the CUDA code
+    float tol = 1e-6; // 	Tolerance to decide if singular or not.
+    int reorder = 0;  // "0" for common matrix without ordering
```

Comment:
const?

Reply:
They will be used as "float" and "int".  If I define them as const here, I will still need to force them to be "float" and "int" again. So Which is better, define "float" directly or define "const float" first and change it to "float" later?

Reply:
why will you need to force them to be non-const later?  I don't see any non-const usage

Reply:
The CUSOLVER function doesn't accept const.

![image](https://user-images.githubusercontent.com/57721574/182201922-0cbfebd2-0585-4e7f-a8f2-cf4b4e6b5dec.png)

Reply:
primitive types are pass by value so it will accept a const as it is just going to copy it